roofit/roofitcore/src/RooRealVar.cxx

```diff
@@ -1311,14 +1316,26 @@ void RooRealVar::installSharedProp(std::shared_ptr<RooRealVarSharedProperties>&&
 /// Stop sharing properties.
 void RooRealVar::deleteSharedProperties()
 {
+  // Nothing to do if there were no shared properties to begin with.
+  if(!_sharedProp) return;
+
+  // Find the std::weak_ptr that the _sharedPropertiesList holds to our
+  // _sharedProp.
+  auto iter = _sharedPropList().find(_sharedProp->asString().Data());
+
+  // Now that we got a reference to the weak_ptr, we can delete our
+  // _sharedProp shared_ptr.
   _sharedProp.reset();
 
-  for (auto it = _sharedPropList().begin(); it != _sharedPropList().end();) {
-    if (it->second.expired()) {
-      it = _sharedPropList().erase(it);
-    } else {
-      ++it;
-    }
+  // If no other RooRealVars shared the shared properties with us, the weak_ptr
+  // in _sharedPropList is now expired and we can erase it from the map. We
+  // also need to check explicitely if the shared property was found in
+  // _sharedPropList to begin with (one might think that this is always the
+  // case by design, but at at the end of a interactive ROOT session it is
+  // might not be the case because _sharedPropList gets deleted before
+  // remaining RooRealVars).
+  if(iter != _sharedPropList().end() && iter->second.expired()) {
+    _sharedPropList().erase(iter);
```

Comment:
I wonder if one shouldn't just erase everything in that list if you find your current `_sharedProp` to be the last instance?
This way, you might also remove all expired pointers that you weren't responsible for (either because of oversights or also in the future if somebody changed the design).

```c++
if (_sharedProp.use_count() <= 1) {
  _sharedProp.reset();
  std::remove_if(_sharedPropList.begin(), _sharedPropList.end(), [](... ptr){ return ptr.expired(); });
  // perform erase
}
```

Reply:
That's a bad strategy, though, if the list is very long, and all items aren't expired yet.

Reply:
I agree, it would be safer to go through the whole list and erase the expired pointers, but I'm worried that this is not good from a performance standpoint. This ``deleteSharedProperties` function is now used in the `RooRealVar` destructor and I don't think it's good to have potentially time consuming logic to it. Especially not for the backports of this PR, as the older releases are expected to behave stable also in performance.